### PR TITLE
Simplify the code to get issue count

### PR DESCRIPTION
**Pull request by lunny**
Since it's a count SQL, we just need to use `Count`.

### CODE REVIEW

**Review by delvh — APPROVED** [review]

**Comment by delvh** [comment]
Wait, does that actually work?
I mean, it compiles, but will the runtime behavior be the same as before?

File: models/issue.go [hunk 1]
```diff
@@ -1510,20 +1510,14 @@ func Issues(opts *IssuesOptions) ([]*Issue, error) {
...
 	opts.setupSessionNoLimit(sess)
-	if err := sess.Find(&countsSlice); err != nil {
+	if _, err := sess.Get(&cnt); err != nil {
 		return 0, fmt.Errorf("unable to CountIssues: %w", err)
 	}
-	if len(countsSlice) != 1 {
-		return 0, fmt.Errorf("unable to get one row result when CountIssues, row count=%d", len(countsSlice))
-	}
-	return countsSlice[0].Count, nil
+	return cnt, nil
```
Comment (delvh) [inline models/issue.go]: If that works, shouldn't
```suggestion
  sess := e.Select("COUNT(issue.id) AS count").Table("issue")
  sess.Join("INNER", "repository", "`issue`.repo_id = `repository`.id")
  opts.setupSessionNoLimit(sess)
  return sess.Count()
```
work as well?

**Comment by wxiaoguang** [comment]
This PR is not ideal and the underlying problem is still not resolved.

* If this PR is correct, then it means the old code `sess.Find(&countsSlice)` is wrong. how can it be? then there are more problems:
    - are all legacy code like this correct?
    - should it be a bug in XORM or MySQL?
* If this PR is incorrect, it only hides the underlying bug and returns the incorrect `0` in some cases, which causes that the problem becomes more difficult to be debugged.

**Comment by lunny** [comment]
> This PR is not ideal and the underlying problem is still not resolved.
> 
> * If this PR is correct, then it means the old code `sess.Find(&countsSlice)` is wrong. how can it be? then there are more problems:
>   
>   * are all legacy code like this correct?
>   * should it be a bug in XORM or MySQL? if yes, should it be fixed in XORM side or should detect the buggy MySQL version?
> * If this PR is incorrect, it only hides the underlying bug and returns the incorrect `0` in some cases, which causes that the problem becomes more difficult to be debugged.

As I said, this PR is not a bug fix. I haven't found the problem of the previous code. 
It's a refactor which make the code more simple and more readability.

**Comment by wxiaoguang** [comment]
Does XORM report an error if `Count` meets 0 row in result? If no, how do you know that if the bug still exists or is resolved after this PR? There will be no error in future even there is still a related bugs.

**Comment by lunny** [comment]
> Does XORM report an error if `Count` meets 0 row in result? If no, how do you know that if the bug still exists or is resolved after this PR? There will be no error in future even there is still a related bugs.

If there is a bug there, it's not related XORM or Gitea, it's an upstream problem. Mysql-driver or mysql server.
As a SQL standard, it should always return 1 record if there is no error.

**Comment by wxiaoguang** [comment]
> > Does XORM report an error if `Count` meets 0 row in result? If no, how do you know that if the bug still exists or is resolved after this PR? There will be no error in future even there is still a related bugs.
> 
> If there is a bug there, it's not related XORM or Gitea, it's an upstream problem. Mysql-driver or mysql server. As a SQL standard, it should always return 1 record if there is no error.

Then should Gitea hide such bug, or report such problem to end users (at least in logs)?

**Review by delvh — APPROVED** [review]

**Comment by lunny** [comment]
> > > Does XORM report an error if `Count` meets 0 row in result? If no, how do you know that if the bug still exists or is resolved after this PR? There will be no error in future even there is still a related bugs.
> > 
> > 
> > If there is a bug there, it's not related XORM or Gitea, it's an upstream problem. Mysql-driver or mysql server. As a SQL standard, it should always return 1 record if there is no error.
> 
> Then should Gitea hide such bug, or report such problem to end users (at least in logs)?

I would like we do the later.

**Comment by wxiaoguang** [comment]
Now we have a good chance to find the root case about the bug that `SELECT COUNT()` returns 0 row.

If this PR gets merged, no way and nobody would continue the debug work.

I prefer to find the root case of the bug first, then take this refactoring.

**Review by wxiaoguang — APPROVED** [review]
The issue (https://github.com/go-gitea/gitea/issues/19368) itself is a MariaDB bug with prepared statements.

https://github.com/go-gitea/gitea/issues/19368#issuecomment-1099808176

To handle all similar problems gracefully in future, some improvements from XORM side:

1. Make XORM's `Count` return an error if the result is 0 row.
2. Make XORM report warnings for SQLs with more than xxxx parameters (eg: 16384), then users/developers can know there is something needed to be refactored, they have enough time to optimize the query before the hard limit comes.

**Comment by wxiaoguang** [comment]
I do not think it's worth to be backported, it's not a bug fix.

**Pull request merged**